common/map_coordinate.h

```diff
@@ -0,0 +1,231 @@
+    const double f;
+};
+
+//----------------------------------------------------------------------------
+// BoBRobotics::MapCoordinate::WGS84
+//----------------------------------------------------------------------------
+//! WGS84 datum - used for GPS data
+struct WGS84
+{
+    static const Ellipsoid ellipsoid;
+};
+const Ellipsoid WGS84::ellipsoid{units::length::meter_t(6378137), units::length::meter_t(6356752.314245), 1.0 / 298.257223563};
+
+//----------------------------------------------------------------------------
+// BoBRobotics::MapCoordinate::OSGB36
+//----------------------------------------------------------------------------
+// OSGB-36 datum - used for UK OS map coordinates
+struct OSGB36
+{
+    static const Transform transform;
+    static const Ellipsoid ellipsoid;
+};
+const Transform OSGB36::transform{units::length::meter_t(-446.448), units::length::meter_t(125.157), units::length::meter_t(-542.060),
+                                  20.4894, units::angle::arcsecond_t(-0.1502), units::angle::arcsecond_t(-0.2470), units::angle::arcsecond_t(-0.8421)};
+const Ellipsoid OSGB36::ellipsoid{units::length::meter_t(6377563.396), units::length::meter_t(6356256.909), 1.0 / 299.3249646};
+
+// An OS coordinate consisting of an 'easting' and 'northing' distance
+typedef std::tuple<units::length::meter_t, units::length::meter_t> OSCoordinate;
+
+//----------------------------------------------------------------------------
+// BoBRobotics::MapCoordinate::Ellipsoid
+//----------------------------------------------------------------------------
+//! A latitude and longitude relative to a particular datum
+template<typename D>
+struct LatLon
+{
+    typedef D Datum;
+
+    const units::angle::degree_t lat;
+    const units::angle::degree_t lon;
+};
+
+//----------------------------------------------------------------------------
+// BoBRobotics::MapCoordinate::Cartesian
+//----------------------------------------------------------------------------
+//! A cartesian coordinate relative to a particular datum
+template<typename D>
+struct Cartesian
+{
+    typedef D Datum;
+
+    const units::length::meter_t x;
+    const units::length::meter_t y;
+    const units::length::meter_t z;
+};
+
+
+// Convert latitude and longitude to cartesian
+// Ported from Javascript Ordnance Survey Grid Reference functions (c) Chris Veness 2005-2017 (MIT Licence)
+template<typename Datum>
+Cartesian<Datum> latLonToCartesian(const LatLon<Datum> &latLon)
+{
+    using namespace units::math;
+    using namespace units::length;
+    using namespace units::literals;
+
+    const meter_t h = 0.0_m; // height above ellipsoid - not currently used
+    const double sinPhi = sin(latLon.lat);
+    const double cosPhi = cos(latLon.lat);
+    const double sinLambda = sin(latLon.lon);
+    const double cosLambda = cos(latLon.lon);
+
+    const double eSq = 2.0 * Datum::ellipsoid.f - Datum::ellipsoid.f * Datum::ellipsoid.f;  // 1st eccentricity squared ≡ (a²-b²)/a²
```

Comment:
Is this definitely right? It's just it's different from the equation in the comment (I can see how eccentricity => ``a`` and ``b``, but you're not doing a division). Maybe another comment is in order if you're doing something non-obvious here.

Reply:
It is, but I'm not sure of the purpose of f at all. It's called the flattening but can be calculated as ``f = (a -b) / a`` https://en.wikipedia.org/wiki/Flattening so I'm not sure why the original javascript had it manually calculated and used inconsistently - I will double check the maths and probably remove f

Reply:
Ok so, from Wikipedia:
![eq1](https://user-images.githubusercontent.com/6793242/51387990-0b0fd080-1b20-11e9-819d-03d8443ea136.gif)
Also from Wikipedia:
![codecogseqn](https://user-images.githubusercontent.com/6793242/51387998-0f3bee00-1b20-11e9-930e-c3abe5f19dad.gif)
And via the power of my rusty algebra:
![codecogseqn 1](https://user-images.githubusercontent.com/6793242/51388009-13680b80-1b20-11e9-8e7a-6f6698417138.gif)

So they're equivalent! What do you think I should do:
1. Leave it but update the comment
2. Get rid of the manually calculated ``f`` term and put everything in terms of ``a`` and ``b``
3. Get rid of the manually calculated ``f`` term and add a method to the struct which calculates it from ``a`` and ``b``?

Reply:
I kinda think 2 as ``latLonToOS`` already does that

Reply:
Yeah I think either 1 or 2. It was just the equation in the comment which threw me a bit.